src/internal/task/task.go

```diff
@@ -67,3 +67,8 @@ type CollectFunc = func(index int64, output *types.Any, _ error) error
 // If an error occurs, then it should be returned.
 // This error will be propagated back to the Doer that created the task.
 type ProcessFunc = func(ctx context.Context, input *types.Any) (output *types.Any, _ error)
+
+type Cache interface {
```

Comment:
I think this and the logic above would be better encapsulated as a `CachedDoer`.  So the base service is unmodified and the CachedDoer wraps an inner doer + a cache.  The etcd service is pretty complicated, and I think it's useful valuable to keep it as simple as possible.

Reply:
Yeah, that is what I started with, but that doesn't work well with the index functionality. Each task received through the channel increments an internal index that maps back to the order the tasks were created. Since most of the caching functionality is handled by the cache implementation, I don't think it adds that much complexity to the Doer, but I am open to alternatives that handle the index issue.

Reply:
The CacheDoer.Do method reads from the channel, checks the cache, if there's a hit, call cb, otherwise send it along to the inner Do implementation.  In the callback, put it in the cache, and call the outer cb.

I guess you do need to track the indexes on the outer level too. You have to track the number of cache hits when you send each item in so that you can map the inner indexes to the outer indexes.  The inner ones will be lower by the number of cache hits.  That probably looks like a `map[int]int` and a `var index, cacheHit int`.  You add to the map when you send, and lookup/remove in the callback.  Increment count every time something comes in and cacheHit every time you skip something.

That does seem more complicated just to isolate the caching.  I'm generally in favor of internal complexity over surface area complexity i.e. requiring consumers to understand an additional type when calling `NewDoer`.  But if you already tried it and didn't like how it came out then I'll defer to you.

Reply:
The in-memory map is a scalability concern and I think the interface becomes more complicated if you need to pass index info to the internal Doer. My thinking is that we will eventually make the interface based on functional options, so you wouldn't really need to understand a cache unless you want to use it (you don't really need to understand it now to not use it). I used to start with an optional parameter for this type of stuff, but I am leaning more towards the explicit setting of the parameter to nil when unused and moving to functional options when more parameters are added over time.

Reply:
I disagree about having functional options in an interface.  The larger the interface the less useful it is, and functional options greatly increase the parameter space of a given method and therefore the size of the interface.

Going to resolve this.  Keeping the cache stuff mixed in is fine.

Reply:
I might actually agree with you there, although I would want to think through the pros / cons a bit more.